### PR TITLE
Adds Support for Graphql ^0.11.0

### LINKED ISSUE #259: Update to accept promise returned from subscribe with graphql 0.11.x
The subscribe function from graphql returns a promise that resolves to an AsyncIterator.
This is different than was in 0.10.5, and thus breaks with an exception in server.ts line 431 with and exception: Cannot read property 'then' of undefined

Normally, this wouldn't be an issue, but the dependencies allow for 0.11.1 to be installed.
    "graphql": "^0.10.1"

When I cloned and installed this repo, it only installed 0.10.5. But, since I added subscriptions to my library after the fact, graphql didn't get downgraded or a warning to that effect didn't occur,  this became a serious pain.

How to reproduce:
Take any existing example and run with graphql 0.11.x and it will fail.

Also, the tests fail in 0.11.1

### PR DESCRIPTION
**Pull request by DxCx**
This PR **adds** support for GraphQL ^0.11.0
^0.10.0 will still work.

**BREAKING CHANGE**: Remove support for `SubscriptionManager`
Related to:
https://github.com/apollographql/graphql-subscriptions/pull/104

TODO:

- [X] Make sure all of the significant new logic is covered by tests
- [X] Rebase your changes on master so that they can be merged easily
- [X] Make sure all tests and linter rules pass
- [x] Update CHANGELOG.md with your change

Closes #259

### CODE REVIEW

**Comment by edvinerikson** [comment]
What's the status of this? No one able to review it? @DxCx

**Comment by DxCx** [comment]
@edvinerikson there is still issue with `graphql-subscriptions` which apollo team needs to decided either remove from dependancies or upgrade as well..
also, as you can see there are older PRs which havn't been reviewed as well.. im waiting as well..

**Comment by edvinerikson** [comment]
Alright, I see. Do you know what the dependency is inside of graphql-subscriptions?

I pinged @stubailo on Slack with the link to my PR and mentioned the low activity. I have not get any reply yet though (20min ago or something).

I really need this change because of the async subscribe function.

**Comment by DxCx** [comment]
only subscription manager interface which i think needs to be just moved inside this package until deprecated or just be deprecated as soon as possible.
but there is no reason of `graphql-subscriptions` to be a dependancy here.

**Comment by edvinerikson** [comment]
I agree. I just migrated away from it when I upgraded. Not sure in which version it got deprecated but making a major bump and removing it shouldn't be too hard(?).

**Comment by stubailo** [comment]
I think @NeoPhi will also take a look soon, let's try to get some next steps before the end of the week.

**Comment by zackify** [comment]
@stubailo anything I can do to help get this out quicker? I was stuck for a long time today until finding this haha

**Comment by NeoPhi** [comment]
I'm in favor of merging this in and having the needed changes to `graphql-subscriptions` made. There has been a lot of confusion about the deprecation warning so adding 0.11 support as part of a version bump on this project and the other one feels like the right next steps.

**Comment by edvinerikson** [comment]
Something I would like to have fixed too is the error handling. Not in this
PR though. I spent quite a few hours trying to debug why it didn't work and
it turns out the library catches its own errors and sends straight to the
client without any way of interception on the sever. E.g. Only thing I had
to go on was "cannot read property .then of undefined" without stack trace.
I didn't even know if it was a server or client error. ¯\_(ツ)_/¯ What's
your thoughts about this?

**Comment by NeoPhi** [comment]
@DxCx Sorry if my comment was unclear. I'm not advocating that `graphql-subscriptions` be kept as a dependency. Instead I'm saying that if we remove the dependency here we should also update `graphql-subscriptions` to remove the deprecated `SubscriptionManager` to avoid any confusion.

**Comment by DxCx** [comment]
Oh, that is absolutely correct. :)

**Comment by DxCx** [comment]
will need a new version of `graphql-subscriptions` so CI won't fail on node 4
but accept for that it looks ready for review to me.

**Comment by NeoPhi** [comment]
@DxCx `graphql-subscriptions` v0.5.0 has been released.

File: package.json [hunk 1]
```diff
@@ -34,16 +32,20 @@
...
+    "@types/graphql": "^0.11.2",
     "@types/lodash": "^4.14.68",
     "@types/mocha": "^2.2.41",
     "@types/node": "^8.0.8",
     "@types/sinon": "^2.3.0",
     "@types/ws": "^3.0.0",
     "chai": "^4.0.2",
-    "graphql": "^0.10.1",
+    "graphql": "^0.11.3",
+    "graphql-subscriptions": "^0.4.4",
```
Comment (dotansimha) [inline package.json]: @DxCx update to latest :)

File: src/adapters/subscription-manager.ts [hunk 2]
```diff
@@ -1,110 +0,0 @@
-import { SubscriptionManager } from 'graphql-subscriptions';
```
Comment (dotansimha) [inline src/adapters/subscription-manager.ts]: 🎉 🎉 🎉 🎉 🎉 🎉 🎉 🎉

**Comment by dotansimha** [comment]
@DxCx looks great. just need to update `graphql-subscriptions` version and fix CI issue. 
Back then we talked about "V1" of the transport, and the goal was to remove the support for the old code. 
If we removed the `SubscriptionManager` support, do you think we need to remove the support for the older protocol as well? @DxCx @Urigo

**Comment by DxCx** [comment]
we don't need to remove it but i think we can remove all deprecated code..
we can do it now or next step will be to clean them up and release `graphql-transport-ws`

File: package.json (hunk 1 above)
Comment (DxCx) [reply package.json]: yeah i was working at that since there was a typing still depended.. now it's pushed =)

**Comment by NeoPhi** [comment]
As this is a breaking change release, I'd be in favor of removing the deprecated code.
I think removal of the old protocol should be part of the switch to graphql-transport-ws as @mistic has thoughts on that.

**Comment by DxCx** [comment]
so do we want to remove all the API and keep the protocol legacy layer for the time being?

**Comment by NeoPhi** [comment]
That would be my recommendation.

**Comment by DxCx** [comment]
@stubailo @Urigo @mistic @dotansimha @martijnwalraven would be great to get some advice in here..

**Comment by stubailo** [comment]
I agree with both of you! Let's go ahead and do it.

**Comment by NeoPhi** [comment]
@DxCx Do you have a preference on next steps? My thinking is that the changes to remove the deprecated API are simple and mirror what is happening with the removal of SubscriptionManager. We can then do a breaking change release of subscriptions-transport-ws that gets this package working with 0.11.x. Then as part of the switch to the new package name the old protocol can be removed.

**Comment by DxCx** [comment]
Yeah lets merge this and wait with releasing ill open another PR cleaning
all deprecated api keeping the legacy protocol support

**Comment by mistic** [comment]
@DxCx @NeoPhi I agree with you! In my opinion we should release now a breaking change version without the `SubscriptionsManager` and the legacy protocol support. And that should be the last `subscriptions-transport-ws` release. Next we should release a new package called `graphql-transport-ws`. What do u think about this?

**Comment by stubailo** [comment]
Hmm, why not release the breaking change as the new package right away?

**Comment by mistic** [comment]
Both ways are valid. However I think that releasing the breaking changes (of things already marked as deprecated for 4 months) in the `subscriptions-transport-ws` as it very last released version can help people to instant migrate to the new package with 0 effort. What do u think @stubailo @NeoPhi @DxCx @dotansimha @Urigo ?

**Comment by DxCx** [comment]
@NeoPhi 
this package is already supported by `apollo-link-ws` and works with apollo-client 2 🎉

**Comment by mistic** [comment]
@DxCx I was suggesting almost the same as u! The only difference is that u prefer to keep the old legacy protocol support on the server side in the last version of `subscriptions-transport-ws` right?

**Comment by mistic** [comment]
@DxCx wasn't this the same I suggest? :D 

In order to allow everyone understand and no doubts remaining what we are suggesting is:

- Release the `0.9.0` version of `subscriptions-transport-ws` with the cleaned api, the legacy protocol and without the `SubscriptionsManager`
- Rename the repository (so we can keep the stars and all the other history) and all the documentation for the new `graphql-transport-ws`.
- Clean the old protocol support, update again the documentation and release the first 0.0.1 version of the `graphql-transport-ws` deprecating the old npm `subscriptions-transport-ws` package with a message to use the new one instead. 

What do u think @DxCx @NeoPhi @dotansimha ?

**Comment by DxCx** [comment]
Agreed.
I just think initial version can be 0.1.0

**Comment by mistic** [comment]
yeah @DxCx it can be 👍 

As soon as the others write their thoughts I can proceed with the agreed changes 😄

**Comment by NeoPhi** [comment]
@mistic Sounds good. Let me know if you'd like me to handle any of that.

**Comment by morajabi** [comment]
Whatever are you doing, please add some notes or guides to upgrade... I wasted a few hours because there's a lot of conflict between documentation of Apollo and this package. Thank you for the great work here.

**Pull request merged**